### PR TITLE
Migrating ComponentizeJS from Walrus to Orca

### PR DESCRIPTION
**Pull request by thesuhas**

### CODE REVIEW

**Comment by guybedford** [comment]
This PR still has some conflicts, which is stopping the CI from running. Once it is rebased it would be great to review.

**Comment by thesuhas** [comment]
Hi,
I've merged with main and updated everything to where it's now passing all the checks!

File: crates/spidermonkey-embedding-splicer/src/lib.rs [hunk 1]
```diff
@@ -407,4 +407,36 @@ impl Guest for SpidermonkeyEmbeddingSplicerComponent {
     }
 }
 
+// Has been added as an easy way to debug the rust code
```
Comment (guybedford) [inline crates/spidermonkey-embedding-splicer/src/lib.rs:410]: Can this either be removed or turned into a more formal debugging test?

File: crates/spidermonkey-embedding-splicer/src/splice.rs [hunk 2]
```diff
@@ -485,8 +458,18 @@ fn synthesize_import_functions(
...
-        func_body.binop_at(table_instr_idx + 2, BinaryOp::I32Add);
+        builder.inject_at(
+            table_instr_idx,
+            InstrumentationMode::Before,
+            Operator::LocalGet {
+                local_index: *arg_idx,
+            },
+        );
+        builder.inject_at(
+            table_instr_idx + 1,
```
Comment (guybedford) [inline crates/spidermonkey-embedding-splicer/src/splice.rs:469]: Do you have a sense of why the offset diff here changed from `+ 2` to `+ 1`?

File: crates/spidermonkey-embedding-splicer/src/splice.rs [hunk 3]
```diff
@@ -500,86 +483,81 @@ fn synthesize_import_functions(
...
-                .collect();
+                .enumerate()
+                .map(|(idx, _)| LocalID::from(idx))
+                .collect(); // Collect the arguments of the function
+
+            let arg_ptr = func.add_local(DataType::I32);
+            let ret_ptr = func.add_local(DataType::I32);
+
+            func.i64_const(-0x5F89E29B87429BD2);
+            func.drop();
```
Comment (guybedford) [inline crates/spidermonkey-embedding-splicer/src/splice.rs:545]: Can these lines be removed?

File: crates/spidermonkey-embedding-splicer/src/stub_wasi.rs [hunk 4]
```diff
@@ -18,41 +20,41 @@ fn stub_import<StubFn>(
...
-        FunctionBuilder::new(&mut module.types, params.as_slice(), results.as_slice());
-    let args = stub(&mut builder.func_body())?;
-    let local_func = builder.local_func(args);
+    // let ty_id = module.types.add(&*params, &*results); Do not need to add a new type as replacing a function with the same type
+    // Pass in Import ID as function ID to preserve location
+    builder.replace_import_in_module(module, iid);
+    // let local_func = builder.local_func(args, iid as FunctionID, ty_id);
 
-    module.funcs.get_mut(fid).kind = FunctionKind::Local(local_func);
+    // module.convert_import_fn_to_local(iid, local_func);
```
Comment (guybedford) [inline crates/spidermonkey-embedding-splicer/src/stub_wasi.rs:52]: It should be okay to just remove these lines.

File: crates/spidermonkey-embedding-splicer/src/stub_wasi.rs [hunk 5]
```diff
@@ -369,10 +368,10 @@ fn stub_stdio(module: &mut Module) -> Result<()> {
     })?;
 
     // (func (param i32 i32 i32 i32) (result i32)))
-    let len_local = module.locals.add(ValType::I32);
     stub_import(module, PREVIEW1, "fd_write", |body| {
-        body.local_get(len_local);
-        Ok(vec![len_local])
+        let len_local: u32 = 3; // Index of the last local
```
Comment (guybedford) [inline crates/spidermonkey-embedding-splicer/src/stub_wasi.rs:372]: Would it make sense to rather write:

```suggestion
        let len_local = LocalID(3); // Index of the last local
```

and use this below?

File: crates/spidermonkey-embedding-splicer/src/stub_wasi.rs [hunk 6]
```diff
@@ -297,30 +294,32 @@ fn stub_random(module: &mut Module) -> Result<()> {
...
     // (func (param i32 i64 i32) (result i32)))
-    let clock_id = module.locals.add(ValType::I32);
-    let precision = module.locals.add(ValType::I64);
-    let time_ptr = module.locals.add(ValType::I32);
     stub_import(module, PREVIEW1, "clock_time_get", |body| {
-        body.local_get(time_ptr);
-        body.local_get(time_ptr);
+        let clock_id: u32 = 0; // First Parameter
+        let precision: u32 = 1; // Second Parameter
+        let time_ptr: u32 = 2; // Third Parameter
```
Comment (guybedford) [inline crates/spidermonkey-embedding-splicer/src/stub_wasi.rs:308]: Can these be declared as local types upfront?

File: crates/spidermonkey-embedding-splicer/src/stub_wasi.rs [hunk 7]
```diff
@@ -160,118 +161,114 @@ fn stub_preview1(module: &mut Module) -> Result<()> {
...
-    let outptr = module.locals.add(ValType::I32);
-    let curptr = module.locals.add(ValType::I32);
     stub_import(
         module,
         "wasi:random/random@0.2.0",
         "get-random-bytes",
         |body| {
+            // let num_bytes = body.add_local(DataType::I64);
+            let num_bytes: u32 = 0; // First parameter
+            let retptr: u32 = 1; // Second parametr
```
Comment (guybedford) [inline crates/spidermonkey-embedding-splicer/src/stub_wasi.rs:205]:
```suggestion
            let num_bytes = LocalId(0); // First parameter
            let retptr = LocalId(1); // Second parametr
```

**Review by guybedford — COMMENTED** [review]
Amazing to see the progress here, thanks, I'll aim to give a final review next week.

**Pull request merged**